File: Random Forest/src/main/java/model/SupervisedLearner.java
package model;// ----------------------------------------------------------------
// The contents of this file are distributed under the CC0 license.
// See http://creativecommons.org/publicdomain/zero/1.0/
// ----------------------------------------------------------------

import static java.lang.System.out;

public abstract class SupervisedLearner
{
	/// Return the name of this learner
	public abstract String name();

	/// Train this supervised learner
	public abstract void train(Matrix features, Matrix labels);

	/// Make a prediction
	abstract void predict(double[] sample, double[] prediction);

	/// Measures the misclassifications with the provided test data
	public int countMisclassifications(Matrix features, Matrix labels)
	{
		if(features.getNumRows() != labels.getNumRows())
			throw new IllegalArgumentException("Mismatching number of getNumRows");
		double[] pred = new double[labels.getNumCols()];
		int mis = 0;
		for(int i = 0; i < features.getNumRows(); i++)
		{
			double[] feat = features.row(i);
			predict(feat, pred);
			double[] lab = labels.row(i);
			for(int j = 0; j < lab.length; j++)
			{
				if(pred[j] != lab[j])
					mis++;
			}
		}
		return mis;
	}
}

File: Random Forest/src/main/kotlin/model/learner/SupervisedLearner.kt
package model.learner

import model.Matrix

abstract class SupervisedLearner {
    // Return the name of this learner
    abstract val name: String

    // Train this supervised learner
    abstract fun train(features: Matrix, labels: Matrix)

    // Make a prediction
    abstract fun predict(receiving: DoubleArray, predicting: DoubleArray)

    // Measures the misclassifications with the provided test data
    fun countMisclassifications(features: Matrix, labels: Matrix): Int {
        if (features.rows != labels.rows)
            throw IllegalArgumentException("Mismatching number of rows")
        val pred = DoubleArray(labels.cols)
        var mis = 0
        for (i in 0 until features.rows) {
            val feat = features.row(i)
            predict(feat, pred)
            val lab = labels.row(i)
            for (j in lab.indices) {
                if (pred[j] != lab[j])
                    mis++
            }
        }
        return mis
    }
}